[`pyupgrade`] Add rules to use PEP 695 generics in classes and functions (`UP046`, `UP047`)

ntBre opened the pull request:
## Summary

<!-- What's the purpose of the change? What does it do, and why? -->
This PR extends our [PEP 695](https://peps.python.org/pep-0695) handling from the type aliases handled by `UP040` to generic function and class parameters, as suggested in the latter two examples from #4617:

```python
# Input
T = TypeVar("T", bound=float)
class A(Generic[T]):
    ...

def f(t: T):
    ...

# Output
class A[T: float]:
    ...

def f[T: float](t: T):
    ...
```

I first implemented this as part of `UP040`, but based on a brief discussion during a very helpful pairing session with @AlexWaygood, I opted to split them into rules separate from `UP040` and then also separate from each other. From a quick look, and based on [this issue](https://github.com/asottile/pyupgrade/issues/836), I'm pretty sure neither of these rules is currently in pyupgrade, so I just took the next available codes, `UP046` and `UP047`.

The last main TODO, noted in the rule file and in the fixture, is to handle generic method parameters not included in the class itself, `S` in this case:

```python
T = TypeVar("T")
S = TypeVar("S")

class Foo(Generic[T]):
    def bar(self, x: T, y: S) -> S: ...
```

but Alex mentioned that that might be okay to leave for a follow-up PR.

I also left a TODO about handling multiple subclasses instead of bailing out when more than one is present. I'm not sure how common that would be, but I can still handle it here, or follow up on that too.

I think this is unrelated to the PR, but when I ran `cargo dev generate-all`, it removed the rule code `PLW0101` from `ruff.schema.json`. It seemed unrelated, so I left that out, but I wanted to mention it just in case.

## Test Plan

New test fixture, `cargo nextest run`

Closes #4617, closes #12542

ntBre commented:
I'm not quite sure what's going wrong with `mkdocs`. I tried pasting in the code it suggested and also adding the code to `KNOWN_FORMATTING_VIOLATIONS`, but running `python scripts/check_docs_formatted.py` locally still reported an issue after each attempt. I'm planning to try again tomorrow morning, but otherwise this should still be ready for review.

I was also mildly surprised to see no changes in the ecosystem checks. Do those run with Python 3.12?

dhruvmanila commented:
> I'm not quite sure what's going wrong with `mkdocs`. I tried pasting in the code it suggested and also adding the rule to `KNOWN_FORMATTING_VIOLATIONS`, but running `python scripts/check_docs_formatted.py` locally still reported an issue after each attempt. I'm planning to try again tomorrow morning, but otherwise this should still be ready for review.

I'm not exactly sure but making the change it suggested seems to work for me 😅 

The script runs on the generated docs which means that you'll have to generate the docs every time you make a change in the Rust code. It's always preferable to include the `--generate-docs` when running this script just to make sure that it's running on the latest docs version.


> I was also mildly surprised to see no changes in the ecosystem checks. Do those run with Python 3.12?

Yes, the ecosystem checks are run on Python 3.12
https://github.com/astral-sh/ruff/blob/c6c434cbc75497dbba65caef892801759aeff6d2/.github/workflows/ci.yaml#L404-L420

https://github.com/astral-sh/ruff/blob/c6c434cbc75497dbba65caef892801759aeff6d2/.github/workflows/ci.yaml#L19-L19

MichaReiser commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695.rs:
```diff
@@ -0,0 +1,218 @@
+//! Shared code for [`use_pep695_type_alias`] (UP040) and [`use_pep695_type_parameter`] (UP046)
```
Did you make any changes to the code in this module or did you copy it over as is?

ntBre replied:
I modified `expr_name_to_type_var` to include the `TypeVar.kind` field, factored out `TypeParam::from` from its call site in `use_pep695_type_alias.rs` and added the `match kind`, and added the `fmt_*` functions. The `Visitor` code is untouched, and so is the `TypeVarRestriction` struct.

Sorry for the confusing mix of changes. I modified it all in-place and then tried splitting it out to separate the rules.

MichaReiser commented on crates/ruff_linter/src/rules/pyupgrade/snapshots/ruff_linter__rules__pyupgrade__tests__UP046.py.snap:
```diff
@@ -0,0 +1,118 @@
+UP046.py:9:7: UP046 [*] Generic class `A` uses `Generic` subclass instead of type parameters
+   |
+ 9 | class A(Generic[T]):
+   |       ^^^^^^^^^^^^^ UP046
```
Should we change the diagnostic range to only the `Generic[T]` instead of the entire class header because the name or other base classes aren't a problem.

ntBre replied:
Good call. I thought this looked nice for the examples, but I think your suggestion makes more sense for longer, more realistic class names! And it will be even more important when we handle multiple base classes.

I wanted to make a similar change for functions, but I think it will require tracking the `range` of the parameter in `TypeVar` and possibly even emitting separate diagnostics for each affected parameter. I could cheat a little for classes because I only handle a single base class for now.

I'm picturing a function like this with non-contiguous, unique type parameters as a problematic case:

```python
def h(p: T, another_param, and_another: U): ...
      ^^^^                 ^^^^^^^^^^^^^^
```

I think that's the ideal kind of diagnostic, but I think it requires separate diagnostics and fixes.

MichaReiser replied:
Hmm yeah that's an interesting case (CC @BurntSushi: While not directly related to diagnostics, it is an interesting problem how we may want multiple diagnostic ranges but a single fix). 

Aren't we tracking the ranges for the fix (by only underlining the type?). although that still doesn't scale to multiple parameters unless we extend the range.

ntBre replied:
Yes, in the generic class case, I'm using the assumption of a single base class to underline the type. This is the `range` used later for the diagnostic:
https://github.com/astral-sh/ruff/blob/55be0d8563c2dbcf4126238f6414efc998dc76e1/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs#L109-L116

Once we add a loop to support multiple base classes, like multiple function parameters, I think we'd need to capture each `range` from the parameters as we loop.
https://github.com/astral-sh/ruff/blob/55be0d8563c2dbcf4126238f6414efc998dc76e1/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs#L209-L221

But I could be misunderstanding or missing an easier way to do it.

MichaReiser commented on crates/ruff_linter/resources/test/fixtures/pyupgrade/UP046.py:
```diff
@@ -0,0 +1,53 @@
+P = ParamSpec("P")
+
+
+class A(Generic[T]):
```
Can we add an example where the class has multiple base classes (I don't know if that's valid)

ntBre replied:
I added a test case showing that multiple base classes are not modified currently and also updated the docs to mention this, as Alex suggested below. I'm happy to work on the implementation here if you prefer, but I was planning to handle this in a follow up.

MichaReiser commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695.rs:
```diff
@@ -0,0 +1,218 @@
+}
+
+impl TypeVar<'_> {
+    /// Format `self` into `s`, where `source` is the whole file, which will be sliced to recover
+    /// the `TypeVarRestriction` values for generic bounds and constraints.
+    fn fmt_into(&self, s: &mut String, source: &str) {
+        match self.kind {
+            TypeVarKind::Var => {}
+            TypeVarKind::Tuple => s.push('*'),
+            TypeVarKind::ParamSpec => s.push_str("**"),
+        }
+        s.push_str(&self.name.id);
+        if let Some(restriction) = &self.restriction {
+            s.push_str(": ");
+            match restriction {
+                TypeVarRestriction::Bound(bound) => {
+                    s.push_str(&source[bound.range()]);
+                }
+                TypeVarRestriction::Constraint(vec) => {
+                    let len = vec.len();
+                    s.push('(');
+                    for (i, v) in vec.iter().enumerate() {
+                        s.push_str(&source[v.range()]);
+                        if i < len - 1 {
+                            s.push_str(", ");
+                        }
+                    }
+                    s.push(')');
+                }
+            }
+        }
+    }
```
I generally prefer to implement `Display` over having `fmt_` methods. It has the advantage that you can call `to_string` if you want to format it to a `String` or that it can be formatted without allocating an extra string, by directly writing into the target string (using `f`)


```suggestion
    impl TypeVar<'_> {
        fn display<'a>(&'a self, source: &'a str) -> DisplayTypeVar<'a> {
            DisplayTypeVar { type_var: self, source }
        }
    }

    struct DisplayTypeVar<'a> {
        type_var: &'a TypeVar<'a>,
        source: &'a str
    }

    impl Display for DisplayTypeVar<'_> {
        fn fmt(...) {
            match self.type_var {
                ...
            }
        }
    }


    /// Format `self` into `s`, where `source` is the whole file, which will be sliced to recover
    /// the `TypeVarRestriction` values for generic bounds and constraints.
    fn fmt_into(&self, s: &mut String, source: &str) {
        match self.kind {
            TypeVarKind::Var => {}
            TypeVarKind::Tuple => s.push('*'),
            TypeVarKind::ParamSpec => s.push_str("**"),
        }
        s.push_str(&self.name.id);
        if let Some(restriction) = &self.restriction {
            s.push_str(": ");
            match restriction {
                TypeVarRestriction::Bound(bound) => {
                    s.push_str(&source[bound.range()]);
                }
                TypeVarRestriction::Constraint(vec) => {
                    let len = vec.len();
                    s.push('(');
                    for (i, v) in vec.iter().enumerate() {
                        s.push_str(&source[v.range()]);
                        if i < len - 1 {
                            s.push_str(", ");
                        }
                    }
                    s.push(')');
                }
            }
        }
    }
```

and similar, add a `DisplayTypeVars` wrapper. You could decide to skip the `DisplayTypeVar` struct and instead make `fmt_into` a method on `DisplayTypeVars` with the signature `fmt_type_var(f, checker.source)`. 

Side node: I think `fmt_into` is sort of confusing because `into` generally suggests that it consumes `self` when used as prefix.

ntBre replied:
Ah great suggestion. I reached for `Display` initially but backed up when I realized I needed the `source` string. The intermediate struct is a good trick to keep in mind, thanks!

MichaReiser reviewed (approved):
Nice, this overall looks good. I didn't review `pep695` in detail because it isn't clear what's copied over and what needs reviewing.

The title suggests to me that UP040 now handles generic functions and classes but reading the summary this doesn't seem to be true? 

What I understand from the summary is that this PR simply introduces a new rule? We should change the title accordingly for a clear changelog entry.

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+    generic_kind: GenericKind,
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+enum GenericKind {
+    GenericClass,
+    GenericFunction,
+}
```
I wonder if these should even be the same rule...? Generic functions and generic classes feel quite distinct.

I don't know how methods play into this, though -- we've obviously left them unimplemented for now, but we might want to try fixing them as a followup extension. We might have to fix methods in the same pass as the class definition...? Not sure...

ntBre replied:
Yeah, I was kind of wondering that myself, especially as I was splitting UP040 and this one. The rule function bodies are just different enough that I don't see an easy way of sharing much more code. Should I separate them?

I agree, not sure about methods either. Maybe they fit with functions if we can just subtract the enclosing class's type parameters from the function code (now I'm realizing we might have a problem with nesting in general?). But I could see it going with the class code, the function code, or being a third rule even.

AlexWaygood replied:
Ah, nesting is a very good point! I think it's fine to not offer a fix for generic functions or classes nested inside other generic functions or classes (they're rare, and complicated to get right!). You can check whether a class or function is nested inside any class or function scopes by using this method on the semantic model to iterate up through all parent statements and checking if any are a `StmtClassDef` or a `StmtFunctionDef`: https://github.com/astral-sh/ruff/blob/73798327c62257c79effedb2f038a5649911e6b7/crates/ruff_python_semantic/src/model.rs#L1243-L1250

We should definitely add tests for nested functions and classes as well!

For methods -- after thinking about it more, I think we might be okay? Whether or not there's also a fix for the class statement being applied by a different rule simultaneously, I think the same fix needs to be applied for the method definition:

```diff
  T = TypeVar("T")
  S = TypeVar("S")

  class Foo(Generic[T]):
-     def bar(self, x: T, y: S) -> S: ...
+     def bar[S](self, x: T, y: S) -> S: ...
```

or, if the class has also been fixed...

```diff
  S = TypeVar("S")

  class Foo[T]:
-     def bar(self, x: T, y: S) -> S: ...
+     def bar[S](self, x: T, y: S) -> S: ...
```

So I think it _should_ be okay for methods to be their own rule -- which, I think, means that it would be good for free functions and classes to be separate rules too!

MichaReiser replied:
Sorry for being late to the party but can you elaborate a bit on the reason why we chose to split the rule? The reason for the split aren't clear to me as a user. What's so different between classes and functions that I'd only want this rule enabled for one or the other? I don't think it should matter how much code the implementations can share -- that would be an us problem and we shouldn't forward it to the users by "overwhelming" them with more rules".

MichaReiser replied:
I saw the format issue. While not ideal, I don't think it's reason enough to split a rule because I consider having to configure two rules more inconvenient than a slightly confusing message in the rules table (which we can fix when we refactor Ruff's rule infrastructure to Red Knot's)

MichaReiser replied:
>  but I don't think the sheer number of rules is one of them

I don't think this is necessarily true, given that users ask for presets, better categorization etc. Users struggle with configuring Ruff and the sheer amount of rules is one big factor causing it. Just speaking for myself, finding if a rule already exists is no easy task (you can try pasting it in the playground but good luck if the rule requires configuration). You have to scroll through a sheer endless list of rules. And I think it's a problem we don't understand well yet because users simply tend to not enable rules because they're overwhelmed (or enable all of them) -- giving them a worse experience overall. That's why I keep pushing back on defaulting to more granular rules. There's a non zero cost. 


Looking at the two rules and reading the documentation I think either works. I've a slight preference to just have one, but considering that we split them out from UP40, maybe it's makes sense to have them separate. I'd probably go with separate rules just to avoid extra work.

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+
+/// UP046
+pub(crate) fn non_pep695_generic_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    if checker.settings.target_version < PythonVersion::Py312 {
```
```suggestion
    // PEP-695 syntax is only available on Python 3.12+
    if checker.settings.target_version < PythonVersion::Py312 {
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+        ..
+    } = class_def;
+
+    // it's a runtime error to mix type_params and Generic, so bail out early if we see existing
+    // type_params
```
ooh, maybe we should add another rule for this, if we don't detect it already with something? We could open an issue for it

```pycon
>>> from typing import TypeVar
>>> T = TypeVar("T")
>>> class Foo[S](Generic[T]): ...
... 
Traceback (most recent call last):
  File "<python-input-13>", line 1, in <module>
    class Foo[S](Generic[T]): ...
  File "<python-input-13>", line 1, in <generic parameters of Foo>
    class Foo[S](Generic[T]): ...
  File "/Users/alexw/.pyenv/versions/3.13.1/lib/python3.13/typing.py", line 1279, in _generic_init_subclass
    raise TypeError(
        "Cannot inherit from Generic[...] multiple times.")
TypeError: Cannot inherit from Generic[...] multiple times.
```

ntBre replied:
Issue created! #15620

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+        return;
+    };
+
+    // TODO(brent) only accept a single, Generic argument for now. I think it should be fine to have
+    // other arguments, but this simplifies the fix just to delete the argument list for now
```
I support this -- it's good to cut scope for the first version of the rule. Could you possibly add a note to the rule's documentation explaining this limitation for now, though?

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+        ..
+    } = function_def;
+
+    // TODO(brent) handle methods, for now return early in a class body. For example, an additional
+    // generic parameter on the method needs to be handled separately from one already on the class
```
Could you add a note to the rule's documentation saying that methods are currently not supported?

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+        visitor.vars
+    };
+
+    // Type variables must be unique; filter while preserving order.
+    let mut type_vars = vars
+        .into_iter()
+        .unique_by(|TypeVar { name, .. }| name.id.as_str())
+        .collect::<Vec<_>>();
```
While this is indeed an error, is it something we should be fixing? It would be just as much an error in the user's original class:

```pycon
>>> class Foo[T, T]: ...
  File "<python-input-14>", line 1
    class Foo[T, T]: ...
                 ^
SyntaxError: duplicate type parameter 'T'
>>> from typing import Generic, TypeVar
>>> T = TypeVar("T")
>>> class Bar(Generic[T, T]): ...
... 
Traceback (most recent call last):
  File "<python-input-16>", line 1, in <module>
    class Bar(Generic[T, T]): ...
              ~~~~~~~^^^^^^
  File "/Users/alexw/.pyenv/versions/3.13.1/lib/python3.13/typing.py", line 432, in inner
    return func(*args, **kwds)
  File "/Users/alexw/.pyenv/versions/3.13.1/lib/python3.13/typing.py", line 1235, in _generic_class_getitem
    raise TypeError(
        f"Parameters to {cls.__name__}[...] must all be unique")
TypeError: Parameters to Generic[...] must all be unique
```

I think instead of filtering out duplicates, we should just bail and not emit a diagnostic if we've found duplicate TypeVars. It will fail at runtime and we can't say for sure what the user wanted to happen.

AlexWaygood reviewed (commented):
Thanks, this is great!

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+/// `contravariant` keywords used by `TypeParam` variables. As such, rewriting a `TypeParam`
+/// variable to an in-line type parameter may change its variance.
+///
+/// Unlike `TypeParam` variables, [PEP 695]-style type parameters cannot be used at runtime. For
+/// example, calling `isinstance(x, T)` with type parameter `T` will raise a `TypeError`. As such,
+/// rewriting a `TypeParam` as a type parameter will cause issues for parameters that are used for
+/// such runtime checks.
```
Oh thanks for the clarification! I copied this from UP040 too and thought it applied here. It must be type-alias-specific again.

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+        return;
+    }
+
+    // generally preserve order, but sort by kind so that the order will be TypeVar...,
+    // TypeVarTuple..., ParamSpec...
+    type_vars.sort_by_key(|tv| tv.kind);
```
I was just trying to match the nice lists from the PEP and `typing` docs :laughing: I agree that keeping the original order makes more sense.

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
````diff
@@ -0,0 +1,237 @@
+///
+/// ## Example
+/// ```python
+/// T = TypeVar("T")
````
Thanks! I combined this with the mkdocs fix

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695.rs:
```diff
@@ -0,0 +1,218 @@
+}
+
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+enum TypeVarKind {
+    Var,
+    Tuple,
+    ParamSpec,
+}
```
Sounds good, I also changed the `Tuple` variant to `TypeVarTuple` to match its PEP/`typing` name too.

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,237 @@
+
+/// ## What it does
+///
+/// Checks for use of `TypeParam`, `TypeParamTuple`, and `ParamSpec` annotations on generic
+/// functions and classes.
```
Sounds good, happy to switch. I used these because they're mentioned in the PEP and in the `typing` documentation for the code we're replacing, and UP040 similarly mentions `TypeAlias` and `TypeAliasType`, but more natural language makes sense to me too.

ntBre replied:
Oops, you're exactly right. These aren't even the PEP/`typing` names. Those are `TypeVar` and `TypeVarTuple` not `TypeParam`. I changed this to 

```rust
/// ## What it does
///
/// Checks for use of standalone type variables and parameter specifications in generic functions
/// and classes.
```

in any case. I also updated both UP046 and the original UP040 docs to use `TypeVar` instead of `TypeParam` in the later variance discussion.

AlexWaygood commented on crates/ruff_linter/resources/test/fixtures/pyupgrade/UP046.py:
```diff
@@ -0,0 +1,63 @@
+from collections.abc import Callable
+from typing import Generic, ParamSpec, TypeVar, TypeVarTuple
+
+T = TypeVar("T", bound=float)
+Ts = TypeVarTuple("Ts")
+P = ParamSpec("P")
```
We have a test that uses a bound TypeVar here -- could you also add a test that uses a constrained TypeVar?

ntBre replied:
Good catch, I added one for classes and one for functions!

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695.rs:
```diff
@@ -0,0 +1,241 @@
+    Constraint(Vec<&'a Expr>),
+}
+
+#[derive(Copy, Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
```
nit: with the latest version of the PR, I think this enum no longer needs the `Ord` and `PartialOrd` implementations (and to me there's no "obvious" order they should always go in, so I'd prefer to remove them)

```suggestion
#[derive(Copy, Clone, Debug, Eq, PartialEq)]
```

ntBre replied:
Good call, I mean to make a comment about the ordering, but you've saved me from that too!

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_alias.rs:
```diff
@@ -27,13 +26,13 @@ use crate::settings::types::PythonVersion;
 /// [PEP 695] uses inferred variance for type parameters, instead of the
-/// `covariant` and `contravariant` keywords used by `TypeParam` variables. As
-/// such, rewriting a `TypeParam` variable to a `type` alias may change its
+/// `covariant` and `contravariant` keywords used by `TypeVar` variables. As
+/// such, rewriting a `TypeVar` variable to a `type` alias may change its
 /// variance.
```
This issue predates your PR, but the wording here doesn't really make sense to me; I'd change it to

```suggestion
/// `covariant` and `contravariant` keywords used by `TypeVar` variables. As
/// such, rewriting a type alias using a PEP-695 `type` statement may change
/// the variance of the alias's type parameters.
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_alias.rs:
```diff
@@ -27,13 +26,13 @@ use crate::settings::types::PythonVersion;
-/// Unlike `TypeParam` variables, [PEP 695]-style `type` aliases cannot be used
+/// Unlike `TypeVar` variables, [PEP 695]-style `type` aliases cannot be used
 /// at runtime. For example, calling `isinstance` on a `type` alias will throw
-/// a `TypeError`. As such, rewriting a `TypeParam` via the `type` keyword will
+/// a `TypeError`. As such, rewriting a `TypeVar` via the `type` keyword will
 /// cause issues for parameters that are used for such runtime checks.
```
Blegh, the wording is again pretty bad here :/ Your docs for your new rule are much better!

The wording here confuses two issues:
1. Replacing an assignment with a type statement
2. Replacing implicit variance over a globally defined `TypeVar` with an explicit type parameter

For generic old-style type aliases that use type arguments, this whole paragraph actually doesn't apply at all, since you can't use `isinstance()` with them even if they use the old syntax:

```pycon
>>> from typing import TypeVar, TypeAlias
>>> T = TypeVar("T")
>>> X: TypeAlias = list[T]
>>> isinstance([], X)
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    isinstance([], X)
    ~~~~~~~~~~^^^^^^^
TypeError: isinstance() argument 2 cannot be a parameterized generic
```

You can only use `isinstance()` on old-style type aliases if they are _not_ generic:

```pycon
>>> from typing import TypeAlias
>>> X: TypeAlias = int
>>> isinstance(42, X)
True
```

I'd probably go with something like for the docs here

```
/// Unlike type aliases that use simple assignments, definitions created using
/// [PEP 695] `type` statements cannot be used as drop-in replacements at
/// runtime for the value on the right-hand side of the statement. This means
/// that while for some simple old-style type aliases you can use them as the
/// second argument to an `isinstance()` call (for example), doing the same
/// with a [PEP 695] `type` statement will always raise `TypeError` at
/// runtime.
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,258 @@
+///
+/// [PEP 695] uses inferred variance for type parameters, instead of the `covariant` and
+/// `contravariant` keywords used by `TypeVar` variables. As such, rewriting a `TypeVar` variable to
+/// an in-line type parameter may change its variance.
```
```suggestion
/// an inline type parameter may change its variance.
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,258 @@
+/// `contravariant` keywords used by `TypeVar` variables. As such, rewriting a `TypeVar` variable to
+/// an in-line type parameter may change its variance.
+///
+/// The rule currently excludes cases where it conceptually should be able to give a diagnostic. In
+/// particular, it skips generic classes with multiple base classes, and it skips generic methods in
+/// generic or non-generic classes.
```
```suggestion
/// The rule currently skips generic classes with multiple base classes, and skips
/// generic methods in generic or non-generic classes.
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
````diff
@@ -0,0 +1,258 @@
+/// T = TypeVar("T")
+///
+///
+/// class GenericClass(Generic[T]): ...
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// class GenericClass[T]: ...
````
Maybe it's worth making the example slightly more realistic by using the type parameter in the class body?

```suggestion
/// class GenericClass(Generic[T]):
///     var: T
/// ```
///
/// Use instead:
///
/// ```python
/// class GenericClass[T]:
///     var: T
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,258 @@
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        let NonPEP695TypeParameter { name, generic_kind } = self;
+        match generic_kind {
+            GenericKind::GenericClass => {
+                format!("Generic class `{name}` uses `Generic` subclass instead of type parameters")
+            }
+            GenericKind::GenericFunction => {
+                format!("Generic function `{name}` should use type parameters")
+            }
+        }
```
Another reason why it might be good for these to be implemented as separate rules: the rules table in the docs [here](https://docs.astral.sh/ruff/rules/) is generated by this method; it will use the contents of the first `format!()` call in this method for this rule's "message" column. Only displaying the first `format!()` call and not the second would give a slightly misleading impression of what the rule does, in this case!

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_parameter.rs:
```diff
@@ -0,0 +1,258 @@
+        visitor.vars
+    };
+
+    // Type variables must be unique; filter while preserving order.
+    let nvars = vars.len();
+    let type_vars = vars
+        .into_iter()
+        .unique_by(|TypeVar { name, .. }| name.id.as_str())
+        .collect::<Vec<_>>();
+
+    if type_vars.is_empty() {
+        return;
+    }
+
+    // non-unique type variables are runtime errors, so just bail out here
+    if type_vars.len() < nvars {
+        return;
+    }
```
nit: to me it makes sense to check the typevars are not empty before filtering out duplicates

```suggestion
    if vars.is_empty() {
        return;
    }

    let type_vars: Vec<&str> = vars
        .iter()
        .unique_by(|tvar| &tvar.name.id)
        .collect();

    // non-unique type variables are runtime errors, so just bail out here
    if type_vars.len() < nvars.len() {
        return;
    }
```

ntBre replied:
This section was bothering me too. I factored it out into a `check_type_vars` function to share it between the class and function rules and applied your suggestion here.

AlexWaygood commented:
Oh, could you also possibly add a mention of https://docs.astral.sh/ruff/rules/unused-private-type-var/ to the docs for this rule? The two rules work very well when they're enabled together, as the fixes for this rule might result in lots of type variables becoming unused. This is how you can reference another rule in the docs for a rule: https://github.com/astral-sh/ruff/blob/4cfa3555199f9377b29c490cbb703beb6b1eb3b4/crates/ruff_linter/src/rules/flake8_type_checking/rules/type_alias_quotes.rs#L116-L119

AlexWaygood commented:
Since we don't have a useful ecosystem report on this PR (due to so few projects setting `requires-python` to Python 3.12 or newer), I tried modifying typeshed's pyproject.toml file to say that typeshed only supported Python 3.12+, and I then ran Ruff using this branch (with only UP046 selected) on my local clone of typeshed. Here's a PR to my typeshed fork with the results: https://github.com/AlexWaygood/typeshed/pull/34.

Summary of the findings:
1. No syntax errors or comments dropped as far as I can see -- it mostly looks fantastic!

2. However, mypy crashes. Given that pyright is basically fine with all the changes here, this is almost certainly a bug in mypy, and probably a typeshed-specific one (typeshed provides the stubs for some very core parts of the standard library, which mypy often makes assumptions about; it's quite easy to provoke a mypy crash by changing random parts of typeshed). Given that it's quite unlikely that this will affect non-typeshed users, I think this probably doesn't matter for us.

3. Pyright _nearly_ passes, but there's a few errors about missing type arguments for `slice`. And this highlights an important point that we both missed! `slice` is defined [here](https://github.com/python/typeshed/blob/57d7c4334b64856fda6f6e8f992b101ddafe2f57/stdlib/builtins.pyi#L943) in typeshed, and we can see it's generic over three type parameters. The type parameters are defined [here](https://github.com/python/typeshed/blob/57d7c4334b64856fda6f6e8f992b101ddafe2f57/stdlib/builtins.pyi#L98-L100) -- and they all have `default` keyword arguments used in their definitions! This is [PEP 696](https://peps.python.org/pep-0696/), which was introduced in Python 3.13. An old-style TypeVar that uses the `default` keyword cannot be rewritten with type parameters on Python 3.12, but if we're on Python 3.13+, we can rewrite it like this:

   ```diff
     from typing_extensions import Any, TypeVar, Generic

     T = TypeVar("T", default=Any, bound=str)

   - class slice(Generic[T]):
   + class slice[T: str = Any]:
         start: T
   ```

4. It's a bit annoying that the rule will rewrite something like this:
   
   ```py
   _T = TypeVar("_T")

   def f(x: _T) -> _T: ...
   ```

   to

   ```py
   def f[_T](x: _T) -> _T: ...
   ```

   when what I'd really _like_ is

   ```py
   def f[T](x: T) -> T: ...
   ```

   The leading underscore is just unnecessary visual noise when it becomes a type parameter,
since type parameters can't anyway be accessed from outside the function scope; they're private by default. I don't think we should try to rename any type variables as part of this PR, but we might consider thinking about this as part of a followup.

ntBre commented:
Wow great catch on the `default` argument. Should we just bail out for now if we see a `default` argument? Eventually I guess we'll need a bit more bookkeeping around the `checker.settings.target_version`.

Otherwise I think the remaining TODOs are 
- [ ] add tests for nested functions and classes (and bail out of nested scopes)
- [ ] split into two rules

Hopefully I've addressed everything else, but please let me know if I missed something!

AlexWaygood commented:
> Should we just bail out for now if we see a `default` argument?

Yup, fine to bail out on seeing a `default` argument for this PR! (Though we should make a note about that in the docs for the rule, or we'll get bug reports about it if we land this and then never get to the followup implementing support for `TypeVar`s that have the `default` argument

ntBre commented:
Okay, I think I covered all of the suggestions! For follow ups I have

- [ ] Rename private type variables in generics (don't make inline generics like `[_T]`)
- [ ] Allow multiple base classes for generic classes
- [ ] Handle generic methods
- [ ] Handle `default` kwargs for Python >= 3.13

Plus the separate, related rule for mixed old- and new-style generics in #15620. Should I open issues for each of these, or just track them here?

MichaReiser commented:
I'd suggest creating a new issue. They all seem like extensions and can be implemented after this PR.

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_function.rs:
```diff
@@ -0,0 +1,157 @@
+/// ## Known problems
+///
+/// [PEP 695] uses inferred variance for type parameters, instead of the `covariant` and
+/// `contravariant` keywords used by `TypeVar` variables. As such, rewriting a `TypeVar` variable to
+/// an inline type parameter may change its variance.
```
```suggestion
/// `contravariant` keywords used by `TypeVar` variables. As such, replacing a `TypeVar` variable
/// with an inline type parameter may change its variance.
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_function.rs:
````diff
@@ -0,0 +1,157 @@
+/// T = TypeVar("T")
+///
+///
+/// def generic_function(t: T):
+///     var: T
+/// ```
+///
+/// Use instead:
+///
+/// ```python
+/// def generic_function[T](t: T):
+///     var: T
````
```suggestion
/// def generic_function(var: T) -> T:
///     return var
/// ```
///
/// Use instead:
///
/// ```python
/// def generic_function[T](var: T) -> T:
///     return var
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_function.rs:
```diff
@@ -0,0 +1,157 @@
+///
+/// ## See also
+///
+/// This rule replaces standalone type variables in class and function signatures but doesn't remove
```
```suggestion
/// This rule replaces standalone type variables in function signatures but doesn't remove
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_function.rs:
```diff
@@ -0,0 +1,157 @@
+///
+/// This rule replaces standalone type variables in class and function signatures but doesn't remove
+/// the corresponding type variables even if they are unused after the fix. See
+/// [`unused-private-type-var`](unused-private-type-var.md) for a rule to clean up unused type
+/// variables.
```
```suggestion
/// [`unused-private-type-var`](unused-private-type-var.md) for a rule to clean up unused
/// private type variables.
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_class.rs:
```diff
@@ -0,0 +1,155 @@
+///
+/// This rule replaces standalone type variables in class and function signatures but doesn't remove
+/// the corresponding type variables even if they are unused after the fix. See
+/// [`unused-private-type-var`](unused-private-type-var.md) for a rule to clean up unused type
+/// variables.
```
```suggestion
/// [`unused-private-type-var`](unused-private-type-var.md) for a rule to clean up unused
/// private type variables.
```

ntBre commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,292 @@
+        return None;
+    }
+
+    Some(type_vars)
```
Oh nice! I always forget about the methods on `bool`.

ntBre commented:
Thanks for all of the documentation fixes and the `then_some` improvement! It looks like there's a tiny typo that I'm fixing locally. Otherwise, do we want to [revisit](https://github.com/astral-sh/ruff/pull/15565#discussion_r1923216394) splitting the rule? If not, I just want to extend the `See also` sections to point to each other. I opened the other issue for follow ups too.

AlexWaygood commented:
I updated the [PR against my typeshed fork](https://github.com/AlexWaygood/typeshed/pull/34) to use the latest version of this PR. The pyright CI there is flagging a couple of other issues:
1. `/home/runner/work/typeshed/typeshed/stdlib/builtins.pyi:1547:32 - error: Type parameter "SupportsRichComparisonT" is not included in the type parameter list for "max" (reportGeneralTypeIssues)`.

  This is because `SupportsRichComparisonT` is defined in another module, so we can't see that it's a type variable. This actually means that the autofix breaks the function definition, because the function now mixes old-style type variable declarations with new-style type parameters in its annotations, and pyright correctly rejects this. I don't think there's a good solution here other than to make sure that the fix is marked unsafe, and to document that this is a known limitation of the rule.

2. In a similar vein, it looks like we don't currently recognise `typing.AnyStr` as being a type variable. This is the same issue as issue (1), but it's a very commonly used type variable (and a public member of the standard-library `typing` module), so I think it's worth us adding some special-casing for it so that it's recognised as a type variable.
3.

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_class.rs:
```diff
@@ -0,0 +1,158 @@
+            },
+            *range,
+        )
+        .with_fix(Fix::applicable_edit(
+            Edit::replacement(type_params.to_string(), name.end(), arguments.end()),
+            Applicability::Safe,
+        )),
```
the fixes to these rules should be marked as unsafe rather than safe:
- The issue pointed out in the `Known issues` section regarding the switch from explicit variance to inferred variance means that they could change the semantics of the code's annotations
- the issue I pointed out just now regarding type variables from other modules means that the fixes could in some cases outright break the code's annotations

```suggestion
        .with_fix(Fix::unsafe_edit(Edit::replacement(
            type_params.to_string(),
            name.end(),
            arguments.end(),
        ))),
```

We also need to add a `## Fix safety` section to the rules' docs explaining why the fixes are marked as unsafe

AlexWaygood replied:
(if you grep for `## Fix safety`, you can find lots of other rules that have similar sections in their docs, that you can use as examples!)

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_generic_function.rs:
```diff
@@ -0,0 +1,160 @@
+            },
+            TextRange::new(name.start(), parameters.end()),
+        )
+        .with_fix(Fix::applicable_edit(
+            Edit::insertion(type_params.to_string(), name.end()),
+            Applicability::Safe,
+        )),
```
```suggestion
        .with_fix(Fix::unsafe_edit(Edit::insertion(
            type_params.to_string(),
            name.end(),
        ))),
```

ntBre commented:
Thanks again for the thorough review and for your `typeshed` tests! I think I have now
* handled imported type variables (by giving only a diagnostic in classes and by giving an unsafe fix (that excludes them) in functions)
* updated the `Known issues` and `Fix safety` sections of both rules
* added a special case for `AnyStr`

The `AnyStr` case felt a little awkward, so hopefully it's at least close to what you had in mind.

AlexWaygood commented on crates/ruff_linter/resources/test/fixtures/pyupgrade/UP046.py:
```diff
@@ -0,0 +1,80 @@
+# little more work. these should be left alone for now but be fixed eventually.
+class NotGeneric:
+    # -> generic_method[T: float](t: T)
+    def generic_method(t: T):
+        pass
```
micro-nit: it's nice to keep fixtures looking like "realistic" Python code where possible (this looks a bit strange to me, as there's generally no reason to use a `TypeVar` in a function's annotations like this unless either two parameters are annotated with the `TypeVar` or the return annotation uses the `TypeVar`):

```suggestion
    def generic_method(var: T) -> T:
        return var
```

this applies to a few other functions and classes in your fixtures, but I won't comment on all of them ;)

ntBre replied:
Ah sorry for making you repeat yourself. I know you mentioned it for the docs, but I thought I could get away with it for the tests! I'll keep this in mind now.

AlexWaygood commented on crates/ruff_linter/resources/test/fixtures/pyupgrade/UP046.py:
```diff
@@ -0,0 +1,80 @@
+
+# This case gets a diagnostic but not a fix because we can't look up the bounds
+# or constraints on the generic type from another module
+class ExternalType(Generic[T, SupportsRichComparisonT]):
```
should `SupportsRichComparisonT` be imported from another module at the top of this file?

ntBre replied:
I didn't know where it's actually from (maybe `_typeshed` from a github search?), but I imported it from `somewhere` at the top to make it more clear that we're testing an unknown type variable,  not just a missing import. Thanks!

AlexWaygood replied:
yup, it's defined in `_typeshed` here: https://github.com/python/typeshed/blob/ce521e8c76eec84828b30353b163fcff8e054bbd/stdlib/_typeshed/__init__.pyi#L104 (it doesn't actually exist at runtime!). But importing it from `somewhere` is fine for these fixtures; the only important thing is that it needs to come from another module 👍

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,335 @@
+enum TypeVarRestriction<'a> {
+    /// A type variable with a bound, e.g., `TypeVar("T", bound=int)`.
+    Bound(&'a Expr),
+    /// A type variable with constraints, e.g., `TypeVar("T", int, str)`.
+    Constraint(Vec<Expr>),
```
Is it worth using a `std::borrow::Cow` here, so that we can avoid cloning in the common non-`AnyStr` case?

```suggestion
    /// A type variable with constraints, e.g., `TypeVar("T", int, str)`.
    Constraint(Vec<Cow<'a, Expr>>),
```

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,335 @@
+                    let len = vec.len();
+                    f.write_str("(")?;
+                    for (i, v) in vec.iter().enumerate() {
+                        // typing.AnyStr special case doesn't have a real range
+                        if let Expr::Name(name) = v {
+                            f.write_str(name.id.as_ref())?;
+                        } else {
+                            f.write_str(&self.source[v.range()])?;
+                        }
```
hrm... I'm not a huge fan of the way that the special handling for `AnyStr` is kind-of implicit here. It also seems a shame that we're now cloning all `TypeVarRestriction::Constraint` `Expr`s just to handle the `AnyStr` special case.

I think we can solve this by:
1. Adding a dedicated variant for `TypeVarRestriction`, and
2. Only storing an `&'a str` for the `name` field on `TypeVar`, rather than a whole `ast::name::Name`.

Something like this? I think this patch also has the advantage that we'd recognise a fully qualified use of `AnyStr` as an attribute expression, e.g. `import typing; x: typing.AnyStr`:

<details>
<summary>Suggested patch</summary>

```diff
diff --git a/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs b/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs
index 3de4228bc..de9a6407a 100644
--- a/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs
+++ b/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs
@@ -28,7 +28,10 @@ enum TypeVarRestriction<'a> {
     /// A type variable with a bound, e.g., `TypeVar("T", bound=int)`.
     Bound(&'a Expr),
     /// A type variable with constraints, e.g., `TypeVar("T", int, str)`.
-    Constraint(Vec<Expr>),
+    Constraint(Vec<&'a Expr>),
+    /// `AnyStr` is a special case: the only public `TypeVar` defined in the standard library,
+    /// and thus the only one that we recognise when imported from another module.
+    AnyStr,
 }
 
 #[derive(Copy, Clone, Debug, Eq, PartialEq)]
@@ -40,7 +43,7 @@ enum TypeParamKind {
 
 #[derive(Debug)]
 struct TypeVar<'a> {
-    name: &'a ExprName,
+    name: &'a str,
     restriction: Option<TypeVarRestriction<'a>>,
     kind: TypeParamKind,
     default: Option<&'a Expr>,
@@ -92,23 +95,19 @@ impl Display for DisplayTypeVar<'_> {
             TypeParamKind::TypeVarTuple => f.write_str("*")?,
             TypeParamKind::ParamSpec => f.write_str("**")?,
         }
-        f.write_str(&self.type_var.name.id)?;
+        f.write_str(self.type_var.name)?;
         if let Some(restriction) = &self.type_var.restriction {
             f.write_str(": ")?;
             match restriction {
                 TypeVarRestriction::Bound(bound) => {
                     f.write_str(&self.source[bound.range()])?;
                 }
+                TypeVarRestriction::AnyStr => f.write_str("(bytes, str)")?,
                 TypeVarRestriction::Constraint(vec) => {
                     let len = vec.len();
                     f.write_str("(")?;
                     for (i, v) in vec.iter().enumerate() {
-                        // typing.AnyStr special case doesn't have a real range
-                        if let Expr::Name(name) = v {
-                            f.write_str(name.id.as_ref())?;
-                        } else {
-                            f.write_str(&self.source[v.range()])?;
-                        }
+                        f.write_str(&self.source[v.range()])?;
                         if i < len - 1 {
                             f.write_str(", ")?;
                         }
@@ -135,7 +134,7 @@ impl<'a> From<&'a TypeVar<'a>> for TypeParam {
             TypeParamKind::TypeVar => {
                 TypeParam::TypeVar(TypeParamTypeVar {
                     range: TextRange::default(),
-                    name: Identifier::new(name.id.clone(), TextRange::default()),
+                    name: Identifier::new(*name, TextRange::default()),
                     bound: match restriction {
                         Some(TypeVarRestriction::Bound(bound)) => Some(Box::new((*bound).clone())),
                         Some(TypeVarRestriction::Constraint(constraints)) => {
@@ -146,6 +145,25 @@ impl<'a> From<&'a TypeVar<'a>> for TypeParam {
                                 parenthesized: true,
                             })))
                         }
+                        Some(TypeVarRestriction::AnyStr) => {
+                            Some(Box::new(Expr::Tuple(ast::ExprTuple {
+                                range: TextRange::default(),
+                                elts: vec![
+                                    Expr::Name(ExprName {
+                                        range: TextRange::default(),
+                                        id: Name::from("str"),
+                                        ctx: ast::ExprContext::Load,
+                                    }),
+                                    Expr::Name(ExprName {
+                                        range: TextRange::default(),
+                                        id: Name::from("bytes"),
+                                        ctx: ast::ExprContext::Load,
+                                    }),
+                                ],
+                                ctx: ast::ExprContext::Load,
+                                parenthesized: true,
+                            })))
+                        }
                         None => None,
                     },
                     // We don't handle defaults here yet. Should perhaps be a different rule since
@@ -155,12 +173,12 @@ impl<'a> From<&'a TypeVar<'a>> for TypeParam {
             }
             TypeParamKind::TypeVarTuple => TypeParam::TypeVarTuple(TypeParamTypeVarTuple {
                 range: TextRange::default(),
-                name: Identifier::new(name.id.clone(), TextRange::default()),
+                name: Identifier::new(*name, TextRange::default()),
                 default: None,
             }),
             TypeParamKind::ParamSpec => TypeParam::ParamSpec(TypeParamParamSpec {
                 range: TextRange::default(),
-                name: Identifier::new(name.id.clone(), TextRange::default()),
+                name: Identifier::new(*name, TextRange::default()),
                 default: None,
             }),
         }
@@ -178,36 +196,27 @@ struct TypeVarReferenceVisitor<'a> {
 /// Recursively collects the names of type variable references present in an expression.
 impl<'a> Visitor<'a> for TypeVarReferenceVisitor<'a> {
     fn visit_expr(&mut self, expr: &'a Expr) {
+        // special case for typing.AnyStr, which is a commonly-imported type variable in the
+        // standard library with the definition:
+        //
+        // ```python
+        // AnyStr = TypeVar('AnyStr', bytes, str)
+        // ```
+        //
+        // As of 01/2025, this line hasn't been modified in 8 years, so hopefully there won't be
+        // much to keep updated here. See
+        // https://github.com/python/cpython/blob/383af395af828f40d9543ee0a8fdc5cc011d43db/Lib/typing.py#L2806
+        if self.semantic.match_typing_expr(expr, "AnyStr") {
+            self.vars.push(TypeVar {
+                name: "AnyStr",
+                restriction: Some(TypeVarRestriction::AnyStr),
+                kind: TypeParamKind::TypeVar,
+                default: None,
+            });
+            return;
+        }
+
         match expr {
-            // special case for typing.AnyStr, which is a commonly-imported type variable in the
-            // standard library with the definition:
-            //
-            // ```python
-            // AnyStr = TypeVar('AnyStr', bytes, str)
-            // ```
-            //
-            // As of 01/2025, this line hasn't been modified in 8 years, so hopefully there won't be
-            // much to keep updated here. See
-            // https://github.com/python/cpython/blob/383af395af828f40d9543ee0a8fdc5cc011d43db/Lib/typing.py#L2806
-            e @ Expr::Name(name) if self.semantic.match_typing_expr(e, "AnyStr") => {
-                self.vars.push(TypeVar {
-                    name,
-                    restriction: Some(TypeVarRestriction::Constraint(vec![
-                        Expr::Name(ExprName {
-                            range: TextRange::default(),
-                            id: Name::from("bytes"),
-                            ctx: ruff_python_ast::ExprContext::Load,
-                        }),
-                        Expr::Name(ExprName {
-                            range: TextRange::default(),
-                            id: Name::from("str"),
-                            ctx: ruff_python_ast::ExprContext::Load,
-                        }),
-                    ])),
-                    kind: TypeParamKind::TypeVar,
-                    default: None,
-                });
-            }
             Expr::Name(name) if name.ctx.is_load() => {
                 if let Some(var) = expr_name_to_type_var(self.semantic, name) {
                     self.vars.push(var);
@@ -243,7 +252,7 @@ fn expr_name_to_type_var<'a>(
         }) => {
             if semantic.match_typing_expr(subscript_value, "TypeVar") {
                 return Some(TypeVar {
-                    name,
+                    name: &name.id,
                     restriction: None,
                     kind: TypeParamKind::TypeVar,
                     default: None,
@@ -288,14 +297,14 @@ fn expr_name_to_type_var<'a>(
                     Some(TypeVarRestriction::Bound(&bound.value))
                 } else if arguments.args.len() > 1 {
                     Some(TypeVarRestriction::Constraint(
-                        arguments.args.iter().skip(1).cloned().collect(),
+                        arguments.args.iter().skip(1).collect(),
                     ))
                 } else {
                     None
                 };
 
                 return Some(TypeVar {
-                    name,
+                    name: &name.id,
                     restriction,
                     kind,
                     default,
@@ -327,7 +336,7 @@ fn check_type_vars(vars: Vec<TypeVar<'_>>) -> Option<Vec<TypeVar<'_>>> {
     // found on the type parameters
     (vars
         .iter()
-        .unique_by(|tvar| &tvar.name.id)
+        .unique_by(|tvar| tvar.name)
         .filter(|tvar| tvar.default.is_none())
         .count()
         == vars.len())
diff --git a/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_alias.rs b/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_alias.rs
index c2340c609..1fb63f8be 100644
--- a/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_alias.rs
+++ b/crates/ruff_linter/src/rules/pyupgrade/rules/pep695/use_pep695_type_alias.rs
@@ -133,7 +133,7 @@ pub(crate) fn non_pep695_type_alias_type(checker: &mut Checker, stmt: &StmtAssig
         .map(|expr| {
             expr.as_name_expr().map(|name| {
                 expr_name_to_type_var(checker.semantic(), name).unwrap_or(TypeVar {
-                    name,
+                    name: &name.id,
                     restriction: None,
                     kind: TypeParamKind::TypeVar,
                     default: None,
@@ -199,7 +199,7 @@ pub(crate) fn non_pep695_type_alias(checker: &mut Checker, stmt: &StmtAnnAssign)
     // Type variables must be unique; filter while preserving order.
     let vars = vars
         .into_iter()
-        .unique_by(|TypeVar { name, .. }| name.id.as_str())
+        .unique_by(|tvar| tvar.name)
         .collect::<Vec<_>>();
 
     checker.diagnostics.push(create_diagnostic(
```

</details>

ntBre replied:
Agreed, I didn't like it either. Thanks for the patch!

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,335 @@
+                        }
+                        None => None,
+                    },
+                    // We don't handle defaults here yet. Should perhaps be a different rule since
+                    // defaults are only valid in 3.13+.
+                    default: None,
```
Should we also add a test for UP040 with a type alias that uses a `TypeVar` with a default (since that rule now shares the logic from this module)?

We might want to see if we can convert the fix there to also use simple string interpolation rather than the `ExprGenerator`... I feel like it makes sense for all three rules to use basically the same infrastructure. But not for this PR! Just another possible followup for you ;)

ntBre replied:
Added some tests and three lines to bail out if we see `default`, and I added the string suggestion to the follow up issue.

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,335 @@
+            e @ Expr::Name(name) if self.semantic.match_typing_expr(e, "AnyStr") => {
+                self.vars.push(TypeVar {
+                    name,
+                    restriction: Some(TypeVarRestriction::Constraint(vec![
+                        Expr::Name(ExprName {
+                            range: TextRange::default(),
+                            id: Name::from("bytes"),
+                            ctx: ruff_python_ast::ExprContext::Load,
+                        }),
+                        Expr::Name(ExprName {
+                            range: TextRange::default(),
+                            id: Name::from("str"),
+                            ctx: ruff_python_ast::ExprContext::Load,
+                        }),
+                    ])),
```
I think we have to make sure here that `str` and `bytes` haven't been overridden in an enclosing scope. E.g. <https://mypy-play.net/?mypy=latest&python=3.12&gist=100810bb27581625b327bce54ec1b63b>. You can do this using the `SemanticModel::is_available()` method.

AlexWaygood replied:
To be clear: I'd _only_ do this for `AnyStr`, at least for the first version of this rule. I think theoretically it could apply to the bounds or constraints for any `TypeVar`, but:
1. It's much more likely to be an issue for a TypeVar coming from another module like `AnyStr`
2. I think it'll just be much more complicated to account for name shadowing with arbitrary non-`AnyStr` TypeVars, and it's probably not essential since the fix is already marked as unsafe.

AlexWaygood replied:
Oh, actually, `SemanticModel::has_builtin_binding` might be better for this use case:

https://github.com/astral-sh/ruff/blob/13e7afca42e1ab4e698412df903e3dd172909caa/crates/ruff_python_semantic/src/model.rs#L268-L277

ntBre replied:
Great catch! I also added a test for this. Is there a way to reset the builtin status of `str`, for example? I tried this initially to check both `str` and `bytes` and that everything was reset, but I think any assignment to a builtin might just mean it's not built in anymore? And we can't nest within a class or function to get a new scope since we skip nesting too! I'm happy enough with the current test, which is just the first case here, but I wanted to check.

```python
str = "string"
class BadStr(Generic[AnyStr]):
    var: AnyStr

str = __builtins__.str

bytes = b"bytes"
class BadBytes(Generic[AnyStr]):
    var: AnyStr

bytes = __builtins__.bytes

class OkayAgain(Generic[AnyStr]):
    var: AnyStr
```

AlexWaygood replied:
> Is there a way to reset the builtin status of `str`, for example?

Ooh good point. Usually I'd usually put everything inside a function but here that obviously doesn't work, for the reasons you pointed out! Maybe you could put the test in a different fixture file entirely (with a docstring at the top of the fixture file saying why it's separated from most of the rest of the fixtures for these rules)?

ntBre replied:
Ah okay, I just put it at the bottom, but a separate file probably makes more sense!

AlexWaygood reviewed (commented):
The docs look fantastic now! Thanks again for your patience!

AlexWaygood commented:
I did some more testing using the PR to my typeshed fork. I have good news and bad news!

The good news is that the pyright report on the typeshed PR is now basically clean -- all remaining errors are due to TypeVars coming from other modules, but we've flagged that in the rule's documentation. Also, these just aren't that common, so there aren't very many errors relating to even that.

The bad news is that I managed to minimize the mypy crash, and it doesn't actually look like something that would only affect typeshed: https://github.com/python/mypy/issues/18507. I think it could potentially affect any stub authors. Pytype is also crashing on the PR to my typeshed fork, so we should perhaps add a note to this rule's documentation saying that not all type checkers fully support PEP 695 yet, so users may need to be careful before applying the rule's autofixes to their codebase.

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,335 @@
+//! Shared code for [`use_pep695_type_alias`] (UP040) and [`use_pep695_type_parameter`] (UP046)
```
And UP047!

AlexWaygood commented on crates/ruff_linter/resources/test/fixtures/pyupgrade/UP046_0.py:
```diff
@@ -0,0 +1,90 @@
+
+
+# This case gets a diagnostic but not a fix because we can't look up the bounds
+# or constraints on the generic type from another module
```
```suggestion
# or constraints on the TypeVar imported from another module
```

AlexWaygood commented on crates/ruff_linter/resources/test/fixtures/pyupgrade/UP046_0.py:
```diff
@@ -0,0 +1,90 @@
+    s: AnyStr
+
+
+class MultipleGenerics(Generic[S, T, Ts, P]):
```
for a `TypeVarTuple`, this would either be

```suggestion
class MultipleGenerics(Generic[S, T, *Ts, P]):
```

or, if the code was originally written to support Python <3.11

```suggestion
class MultipleGenerics(Generic[S, T, Unpack[Ts], P]):
```

with `Unpack` imported from `typing` or `typing_extensions`. Proper handling of `Unpack` is yet another thing you could possibly handle in a followup PR 😅

ntBre replied:
Ah, I thought I caught all of these this time. Thanks!

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,352 @@
+//! Shared code for [`use_pep695_type_alias`] (UP040),
+//! [`use_pep695_generic_class`] (UP046), and [`use_pep695_generic_function`]
+//! (UP047)
```
Hmm... I suppose these new rules don't really fit with our [naming convention](https://github.com/astral-sh/ruff/blob/main/CONTRIBUTING.md#rule-naming-convention). On the other hand, they do match the name of UP040, which is a very similar pre-existing rule. Not sure whether we should use our "general convention" or the "local convention" set by UP040... @MichaReiser, do you have any preference?

ntBre replied:
I think the rule names are actually okay, but the modules don't match the rule names :sweat_smile: I think that should be an easy and non-user-facing change if we wanted to rename all three?

https://github.com/astral-sh/ruff/blob/9fadf46933bb5cc7f6eafbcd22418dfdd8021e3a/crates/ruff_linter/src/codes.rs#L537

https://github.com/astral-sh/ruff/blob/9fadf46933bb5cc7f6eafbcd22418dfdd8021e3a/crates/ruff_linter/src/codes.rs#L543-L544

AlexWaygood replied:
Ohh perfect. Yes, let's just rename the files to match the rules that are in them, then 😄

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,352 @@
+                        }
+                        None => None,
+                    },
+                    // We don't handle defaults here yet. Should perhaps be a different rule since
+                    // defaults are only valid in 3.13+.
```
(I'm not sure it should be a different _rule_ but let's discuss that on https://github.com/astral-sh/ruff/issues/15642 ;)

ntBre replied:
I'll blame Zanie for that one since it was already there! But sounds good to discuss in any case :)

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,352 @@
+    semantic: &'a SemanticModel,
+    name: &'a ExprName,
+) -> Option<TypeVar<'a>> {
+    let Some(Stmt::Assign(StmtAssign { value, .. })) = semantic
+        .lookup_symbol(name.id.as_str())
+        .and_then(|binding_id| {
+            semantic
+                .binding(binding_id)
+                .source
+                .map(|node_id| semantic.statement(node_id))
+        })
+    else {
+        return None;
+    };
```
I know this is just copied over from where it was before -- but we can simplify this quite a bit with some `?`s:

```suggestion
    let StmtAssign { value, .. } = semantic
        .lookup_symbol(name.id.as_str())
        .and_then(|binding_id| semantic.binding(binding_id).source)
        .map(|node_id| semantic.statement(node_id))?
        .as_assign_stmt()?;
```

AlexWaygood reviewed (approved):
Brilliant work. This looks ready to land now!

AlexWaygood commented on crates/ruff_linter/src/rules/pyupgrade/rules/pep695/mod.rs:
```diff
@@ -0,0 +1,346 @@
+//! Shared code for [`use_pep695_type_alias`] (UP040),
+//! [`use_pep695_generic_class`] (UP046), and [`use_pep695_generic_function`]
```
```suggestion
//! Shared code for [`non_pep695_type_alias`] (UP040),
//! [`non_pep695_generic_class`] (UP046), and [`non_pep695_generic_function`]
```

ntBre replied:
Ahh, the one place my LSP couldn't rename for me. I did a quick grep this time, and there's also a pyflakes test called `use_pep695_type_aliass` (double `s` included). Should I rename that as well?

ntBre replied:
Sounds good! I'll push this fix and then move towards merging.
